tests/test_hdcache.cpp

```diff
@@ -0,0 +1,172 @@
+        T tmp = ((T)rand()) / INT_MAX;
+        tmp += (k + 1) * num_clusters;
+        if (j % 2 == 0) {
+          tmp += offset;
+        }
+        data[j * num_clusters + k] = tmp;
+      }
+    }
+    v_data.push_back(data);
+  }
+  return std::move(v_data);
+}
+
+template <typename T>
+void print_vectors(std::vector<T *> &vec, int num_elements, int num_clusters)
+{
+  for (int i = 0; i < num_elements; i++) {
+    for (int c = 0; c < num_clusters; c++) {
+      for (auto v : vec) {
+        std::cout << v[i * num_clusters + c] << ":";
+      }
+      std::cout << std::endl;
+    }
+  }
+}
+
+
+bool validate(const int num_clusters, const int elements, bool *predicates)
+{
+  bool res = true;
+  for (int j = 0; j < elements; j++) {
+    // Generate a value for every cluster center
+    for (int k = 0; k < num_clusters; k++) {
+      if (j % 2 == 0 && predicates[j * num_clusters + k] == true) {
+        res = false;
+      } else if (j % 2 == 1 && predicates[j * num_clusters + k] == false) {
+        res = false;
+      }
+    }
+  }
+  return res;
+}
+
+template <typename T>
+bool do_faiss(std::shared_ptr<HDCache<T>> &index,
+              AMSResourceType resource,
+              int nClusters,
+              int nDims,
+              int nElements,
+              float threshold)
+{
+
+  std::vector<const T *> orig_data =
+      generate_vectors<T>(nClusters, nElements, nDims);
+  std::vector<const T *> data = orig_data;
+
+  bool *predicates =
+      ams::ResourceManager::allocate<bool>(nClusters * nElements, resource);
+
+  if (resource == AMSResourceType::DEVICE) {
+    for (int i = 0; i < orig_data.size(); i++) {
+      T *d_data =
+          ams::ResourceManager::allocate<T>(nClusters * nElements, resource);
+      ams::ResourceManager::copy(const_cast<T *>(orig_data[i]),
+                                 d_data,
+                                 nClusters * nElements * sizeof(T));
+      data[i] = d_data;
+    }
+  }
+
+
+  index->evaluate(nClusters * nElements, data, predicates);
+
+  bool *h_predicates = predicates;
+
+  if (resource == AMSResourceType::DEVICE) {
+    h_predicates = ams::ResourceManager::allocate<bool>(nClusters * nElements,
+                                                        AMSResourceType::HOST);
+    ams::ResourceManager::copy(predicates, h_predicates, nClusters * nElements);
+    for (auto d : data) {
+      ams::ResourceManager::deallocate(const_cast<T *>(d),
+                                       AMSResourceType::DEVICE);
+    }
+    ams::ResourceManager::deallocate(predicates, AMSResourceType::DEVICE);
+  }
+
+
+  for (auto h_d : orig_data)
+    ams::ResourceManager::deallocate(const_cast<T *>(h_d),
+                                     AMSResourceType::HOST);
+
+  bool res = validate(nClusters, nElements, h_predicates);
+
+  ams::ResourceManager::deallocate(h_predicates, AMSResourceType::HOST);
+  return res;
+}
+
+
+int main(int argc, char *argv[])
+{
+  using namespace ams;
```

Comment:
Abort if there are missing arguments and print usage.

Reply:
Done